text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
+### TypeScript to the rescue
+The CDK uses TypeScript, and the TypeScript compiler actually protects against this issue. If we actually typed the *tablify* function as
+```ts
+import { Paddable } from 'leftpad';
+function tablify(rows:* *Array<Array<Paddable>>) { 
+  ...
+}
+```
+The TypeScript compiler would correctly resolve that type to Array<Array<Paddable@2>>, and refuse to compile the calling of *tablify* with an argument of type Array<Array<Paddable@1>>, because of the detected incompatibility between the types.
+_TypeScript will refuse to compile if types of multiple copies of a package are mixed(*)._
+> (*) This is true for classes, where it will do nominal typing. For interfaces, TypeScript will do structural typing, which means any object that has members matching an interface definition is considered to implement that interface, whether it has been declared to implement it or not, and so a class from one copy of a library may be considered to implement an interface from a different copy.
+Here is an example for a compiler error caused by a dependency mix:
+```
+Unable to compile TypeScript: bin/app.ts(10,36): error TS2345:
+Argument of type 'import("node_modules/@aws-cdk/core/lib/app").App' is not assignable to parameter of type 'import("node_modules/my-module/node_modules/@aws-cdk/core/lib/app").App'. 
+Types have separate declarations of a private property '_assembly'.
+```
+### Implications on CDK
+This situation arises the following use cases:
+* User wants to use an older version than the latest published one (rollback).
+* User is using a 3rd-party construct library.
+Let’s look at these in turn.
+#### User wants to use an old version
+The first one, colloquially called “rollback” is easy to understand and seemingly has a simple solution, so let’s look at it first. This situation comes up when the user is in one of 2 situations:
+* CDK team has just (accidentally) released a buggy version and they want to roll back to an older version.
+* CDK team has just released a breaking change (valid in experimental modules) and the user is not prepared to invest the effort yet to migrate, so they want to stay at an older version.
+For an example, let’s say CDK has just released version *1.13.0* but the user wants to stay at *1.12.0*. The only way they have to achieve this is to control the versions in their own app’s *package.json*, so they write:
+```json
+"dependencies": {
+  "@aws-cdk/aws-ecs": "1.12.0",
+  "@aws-cdk/core": "1.12.0",
+}
+```
+The lack of a caret makes it a fixed version, indicating they really want *1.12.0* and not “at least *1.12.0*” (which would resolve to *1.13.0* in this situation). However, because of the transitive caret dependency the complete dependency graph would look like this:
+```
+app@1.0.0      -> aws-ecs==1.12.0, core==1.12.0
+aws-ecs@1.12.0 -> core^1.12.0
+=== { core@1.13.0 is available } ===> 
+app
++-- core@1.12.0
++-- aws-ecs@1.12.0
+    +-- core@1.13.0
```

Comment:
This should be a peer dependency, which would not cause NPM to install it transitively.

Reply:
Removed this section.